- mancano dei video da scaricare (8919 clip presenti ma 9006 videoID nel database)
- controllare clip dove non sono stati estratti dei frame

In [1]:
import pandas as pd
import json

# Shot frame extraction
Ci sono 8919 clip di tiro di cui 8034 hanno frame estratti

In [23]:
# Carica il JSON da file
with open('/home/diego/Documents/GitHub/NBA/Computer_vision/shot_frame_extraction_results.json', 'r') as file:
    data = json.load(file)

# Estrai la sezione dei risultati per video
risultati = data['risultati_per_video']

# Converti in DataFrame
df = pd.DataFrame(list(risultati.items()), columns=['nome_video', 'valore'])

len(df)

8919

In [3]:
df[df['valore'] > 0]

nome_video  valore
0     bos-vs-nyk-0022300065_1_11:48.mp4       2
1     bos-vs-nyk-0022300065_1_11:10.mp4       2
2     bos-vs-nyk-0022300065_1_10:42.mp4       2
3     bos-vs-nyk-0022300065_1_10:05.mp4       2
4     bos-vs-nyk-0022300065_1_09:43.mp4       2
...                                 ...     ...
8913  dal-vs-bos-0042300405_4_02:46.mp4       2
8914  dal-vs-bos-0042300405_4_02:25.mp4       2
8915  dal-vs-bos-0042300405_4_02:19.mp4       2
8916  dal-vs-bos-0042300405_4_01:38.mp4       2
8917  dal-vs-bos-0042300405_4_00:48.mp4       2

[8034 rows x 2 columns]

# Database boston
Ho indicate 9006 clip nel database di cui 8946 hanno le coordinate di tiro

In [4]:
with open('/home/diego/Documents/GitHub/NBA/database_bos.json', 'r') as f:
    lines = [json.loads(line) for line in f]

df_bos = pd.DataFrame(lines)

In [5]:
df_bos[df_bos['videoID'].notna()]

gameID   time  quarter  is_home_team action_type  \
0      bos-vs-nyk-0022300065  11:48        1         False        Shot   
2      bos-vs-nyk-0022300065  11:10        1         False        Shot   
4      bos-vs-nyk-0022300065  10:42        1         False        Shot   
7      bos-vs-nyk-0022300065  10:05        1         False        Shot   
9      bos-vs-nyk-0022300065  09:43        1         False        Shot   
...                      ...    ...      ...           ...         ...   
23330  dal-vs-bos-0042300405  02:25        4          True        Shot   
23332  dal-vs-bos-0042300405  02:19        4          True        Shot   
23335  dal-vs-bos-0042300405  01:38        4          True        Shot   
23340  dal-vs-bos-0042300405  00:48        4          True        Shot   
23345  dal-vs-bos-0042300405  00:22        4          True        Shot   

        player  made            shot_type  player_points assist_player  ...  \
0         None   0.0                 None            NaN          None  ...   
2        Brown   0.0            Jump Shot            NaN          None  ...   
4        Tatum   1.0       Jump Bank Shot            2.0          None  ...   
7         None   1.0                 None            NaN         White  ...   
9        Brown   0.0            Jump Shot            NaN          None  ...   
...        ...   ...                  ...            ...           ...  ...   
23330    Brown   0.0        Driving Layup            NaN          None  ...   
23332    White   0.0     Pullup Jump Shot            NaN          None  ...   
23335    Tatum   0.0  Step Back Jump Shot            NaN          None  ...   
23340    White   0.0     Pullup Jump Shot            NaN          None  ...   
23345   Hauser   0.0            Jump Shot            NaN          None  ...   

             Ph4       Ph5    Pa1      Pa2      Pa3     Pa4        Pa5  \
0         Randle  Robinson  White    Brown  Holiday   Tatum  Porzingis   
2         Randle  Robinson  White    Brown  Holiday   Tatum  Porzingis   
4         Randle  Robinson  White    Brown  Holiday   Tatum  Porzingis   
7         Randle  Robinson  White    Brown  Holiday   Tatum  Porzingis   
9         Randle  Robinson  White    Brown  Holiday   Tatum  Porzingis   
...          ...       ...    ...      ...      ...     ...        ...   
23330  Pritchard     Brown  Hardy  Prosper   Lawson  Powell      Green   
23332  Pritchard     Brown  Hardy  Prosper   Lawson  Powell      Green   
23335  Pritchard     Brown  Hardy  Prosper   Lawson  Powell      Green   
23340  Pritchard     Brown  Hardy  Prosper   Lawson  Powell      Green   
23345  Pritchard     Brown  Hardy  Prosper   Lawson  Powell      Green   

      x_coord y_coord                            videoID  
0        23.0    42.0  bos-vs-nyk-0022300065_1_11:48.mp4  
2       348.0   281.0  bos-vs-nyk-0022300065_1_11:10.mp4  
4       229.0    70.0  bos-vs-nyk-0022300065_1_10:42.mp4  
7       252.0    63.0  bos-vs-nyk-0022300065_1_10:05.mp4  
9        89.0   242.0  bos-vs-nyk-0022300065_1_09:43.mp4  
...       ...     ...                                ...  
23330   276.0    81.0  dal-vs-bos-0042300405_4_02:25.mp4  
23332   103.0   276.0  dal-vs-bos-0042300405_4_02:19.mp4  
23335   201.0   307.0  dal-vs-bos-0042300405_4_01:38.mp4  
23340   380.0   211.0  dal-vs-bos-0042300405_4_00:48.mp4  
23345   150.0   301.0  dal-vs-bos-0042300405_4_00:22.mp4  

[9006 rows x 30 columns]

In [11]:
len(df_bos[(df_bos['videoID'].notna()) & (df_bos['action_type'] == "Shot")])

9006

In [6]:
len(df_bos[(df_bos['x_coord'] >= 0) & (df_bos['y_coord'] >= 0)])

8946

In [22]:
len(df_bos[(df_bos['action_type'] == "Shot")])

9006

# Video_bos
Ci sono 17986 video clip delle partite dei boston, ma considera anche i tiri avversari.
Nella creazione del database sono stati collegati solo i video dei bos.

Quindi non in teoria non ci interessano più analizzare questa cartella in quanto i dati delle clip sono gia presenti nel database

In [26]:
import os

# Percorso della cartella
folder_path = 'Video_bos'

# Filtra solo i file video (ad esempio .mp4)
video_files = [f for f in os.listdir(folder_path) if f.endswith('.mp4')]

# Conta i file
print(f"Numero di video nella cartella '{folder_path}':", len(video_files))


Numero di video nella cartella 'Video_bos': 17986


# Database coordinate di tiro
Ci sono 18123 coordinate di tiro ma come per le clip nel merge del database sono stati presi in considerazione solo i dati dei Boston.

Quindi non ci interessano più (dati già nel database principale)

In [27]:
with open('Scraping_data/shot_chart_bos.json', 'r') as f:
    shot_chart = json.load(f)

df_shot = pd.DataFrame(shot_chart)
df_shot['action_type'] = 'Shot'
df_shot

gameID  timer  quarter  x_coord  y_coord  is_home_team  \
0      bos-vs-nyk-0022300065  11:48        1       23       42         False   
1      bos-vs-nyk-0022300065  11:10        1      348      281         False   
2      bos-vs-nyk-0022300065  10:42        1      229       70         False   
3      bos-vs-nyk-0022300065  10:05        1      252       63         False   
4      bos-vs-nyk-0022300065  09:43        1       89      242         False   
...                      ...    ...      ...      ...      ...           ...   
18118  dal-vs-bos-0042300405  02:25        4      276       81          True   
18119  dal-vs-bos-0042300405  02:19        4      103      276          True   
18120  dal-vs-bos-0042300405  01:38        4      201      307          True   
18121  dal-vs-bos-0042300405  00:48        4      380      211          True   
18122  dal-vs-bos-0042300405  00:22        4      150      301          True   

      action_type  
0            Shot  
1            Shot  
2            Shot  
3            Shot  
4            Shot  
...           ...  
18118        Shot  
18119        Shot  
18120        Shot  
18121        Shot  
18122        Shot  

[18123 rows x 7 columns]

# Cose da fare

***Queste sono le azioni di tiro senza coordinate***

In [ ]:
df_bos[df_bos['x_coord'].isna() & df_bos['y_coord'].isna() & df_bos['videoID'].notna()]

gameID   time  quarter  is_home_team action_type  \
1506   bos-vs-min-0022300154  04:36        5         False        Shot   
1508   bos-vs-min-0022300154  04:29        5         False        Shot   
1515   bos-vs-min-0022300154  02:14        5         False        Shot   
1516   bos-vs-min-0022300154  01:45        5         False        Shot   
1517   bos-vs-min-0022300154  01:20        5         False        Shot   
1522   bos-vs-min-0022300154  00:36        5         False        Shot   
1525   bos-vs-min-0022300154  00:13        5         False        Shot   
3317   bos-vs-cha-0022300217  04:43        5         False        Shot   
3318   bos-vs-cha-0022300217  04:10        5         False        Shot   
3323   bos-vs-cha-0022300217  03:18        5         False        Shot   
3326   bos-vs-cha-0022300217  02:38        5         False        Shot   
3328   bos-vs-cha-0022300217  02:33        5         False        Shot   
3329   bos-vs-cha-0022300217  01:40        5         False        Shot   
3333   bos-vs-cha-0022300217  01:04        5         False        Shot   
3335   bos-vs-cha-0022300217  00:58        5         False        Shot   
3336   bos-vs-cha-0022300217  00:30        5         False        Shot   
6191   bos-vs-gsw-0022300362  04:25        5         False        Shot   
6193   bos-vs-gsw-0022300362  04:06        5         False        Shot   
6196   bos-vs-gsw-0022300362  03:04        5         False        Shot   
6197   bos-vs-gsw-0022300362  02:27        5         False        Shot   
6199   bos-vs-gsw-0022300362  02:01        5         False        Shot   
6201   bos-vs-gsw-0022300362  01:34        5         False        Shot   
6204   bos-vs-gsw-0022300362  00:54        5         False        Shot   
6206   bos-vs-gsw-0022300362  00:52        5         False        Shot   
6208   bos-vs-gsw-0022300362  00:35        5         False        Shot   
6210   bos-vs-gsw-0022300362  00:08        5         False        Shot   
6213   bos-vs-gsw-0022300362  00:02        5         False        Shot   
7151   det-vs-bos-0022300421  04:44        5          True        Shot   
7153   det-vs-bos-0022300421  03:42        5          True        Shot   
7155   det-vs-bos-0022300421  03:41        5          True        Shot   
7157   det-vs-bos-0022300421  02:46        5          True        Shot   
7159   det-vs-bos-0022300421  02:13        5          True        Shot   
7162   det-vs-bos-0022300421  01:46        5          True        Shot   
7164   det-vs-bos-0022300421  01:16        5          True        Shot   
7166   det-vs-bos-0022300421  01:09        5          True        Shot   
7168   det-vs-bos-0022300421  00:45        5          True        Shot   
8898   min-vs-bos-0022300517  04:23        5          True        Shot   
8901   min-vs-bos-0022300517  04:06        5          True        Shot   
8902   min-vs-bos-0022300517  03:32        5          True        Shot   
8909   min-vs-bos-0022300517  02:25        5          True        Shot   
8911   min-vs-bos-0022300517  02:15        5          True        Shot   
8916   min-vs-bos-0022300517  01:04        5          True        Shot   
8918   min-vs-bos-0022300517  00:51        5          True        Shot   
17109  bos-vs-atl-0022301060  04:27        5         False        Shot   
17110  bos-vs-atl-0022301060  03:50        5         False        Shot   
17112  bos-vs-atl-0022301060  03:44        5         False        Shot   
17114  bos-vs-atl-0022301060  02:56        5         False        Shot   
17117  bos-vs-atl-0022301060  02:21        5         False        Shot   
17122  bos-vs-atl-0022301060  01:04        5         False        Shot   
17126  bos-vs-atl-0022301060  00:07        5         False        Shot   
21535  ind-vs-bos-0042300301  04:37        5          True        Shot   
21537  ind-vs-bos-0042300301  04:05        5          True        Shot   
21542  ind-vs-bos-0042300301  03:24        5          True        Shot   
21544  ind-vs-bos-00423003

***Analisi Video scaricati e elaborati***

In [24]:
import os
import pandas as pd
import json

# Percorso alla cartella dei video scaricati
video_folder = '/home/diego/Documents/GitHub/NBA/Video_bos'
downloaded_videos = set(os.listdir(video_folder))

# Caricamento del DataFrame
# df_bos = pd.read_csv(...)  # Se non è già caricato

# Aggiunta colonna 'download': 1 se scaricato, 0 se no
df_bos['download'] = df_bos['videoID'].apply(lambda x: 1 if x in downloaded_videos else 0)

# Caricamento file JSON con i risultati dell'elaborazione
with open('Computer_vision/shot_frame_extraction_results.json', 'r') as f:
    extraction_data = json.load(f)['risultati_per_video']

# Aggiunta colonna 'extracted'
def check_extraction(video_id):
    if video_id in extraction_data:
        return 1 if extraction_data[video_id] > 0 else 0
    else:
        return None  # Non presente nei risultati

df_bos['extracted'] = df_bos['videoID'].apply(check_extraction)


In [25]:
df_bos

gameID   time  quarter  is_home_team action_type  \
0      bos-vs-nyk-0022300065  11:48        1         False        Shot   
1      bos-vs-nyk-0022300065  11:31        1         False     Rebound   
2      bos-vs-nyk-0022300065  11:10        1         False        Shot   
3      bos-vs-nyk-0022300065  10:52        1         False     Rebound   
4      bos-vs-nyk-0022300065  10:42        1         False        Shot   
...                      ...    ...      ...           ...         ...   
23341  dal-vs-bos-0042300405  00:39        4          True        Foul   
23342  dal-vs-bos-0042300405  00:39        4          True         SUB   
23343  dal-vs-bos-0042300405  00:39        4          True         SUB   
23344  dal-vs-bos-0042300405  00:33        4          True     Rebound   
23345  dal-vs-bos-0042300405  00:22        4          True        Shot   

        player  made       shot_type  player_points assist_player  ...    Pa1  \
0         None   0.0            None            NaN          None  ...  White   
1         None   NaN            None            NaN          None  ...  White   
2        Brown   0.0       Jump Shot            NaN          None  ...  White   
3         None   NaN            None            NaN          None  ...  White   
4        Tatum   1.0  Jump Bank Shot            2.0          None  ...  White   
...        ...   ...             ...            ...           ...  ...    ...   
23341     None   NaN            None            NaN          None  ...  Hardy   
23342     None   NaN            None            NaN          None  ...  Hardy   
23343     None   NaN            None            NaN          None  ...  Hardy   
23344     None   NaN            None            NaN          None  ...  Hardy   
23345   Hauser   0.0       Jump Shot            NaN          None  ...  Hardy   

           Pa2      Pa3     Pa4        Pa5  x_coord  y_coord  \
0        Brown  Holiday   Tatum  Porzingis     23.0     42.0   
1        Brown  Holiday   Tatum  Porzingis      NaN      NaN   
2        Brown  Holiday   Tatum  Porzingis    348.0    281.0   
3        Brown  Holiday   Tatum  Porzingis      NaN      NaN   
4        Brown  Holiday   Tatum  Porzingis    229.0     70.0   
...        ...      ...     ...        ...      ...      ...   
23341  Prosper   Lawson  Powell      Green      NaN      NaN   
23342  Prosper   Lawson  Powell      Green      NaN      NaN   
23343  Prosper   Lawson  Powell      Green      NaN      NaN   
23344  Prosper   Lawson  Powell      Green      NaN      NaN   
23345  Prosper   Lawson  Powell      Green    150.0    301.0   

                                 videoID download extracted  
0      bos-vs-nyk-0022300065_1_11:48.mp4        1       1.0  
1                                   None        0       NaN  
2      bos-vs-nyk-0022300065_1_11:10.mp4        1       1.0  
3                                   None        0       NaN  
4      bos-vs-nyk-0022300065_1_10:42.mp4        1       1.0  
...                                  ...      ...       ...  
23341                               None        0       NaN  
23342                               None        0       NaN  
23343                               None        0       NaN  
23344                               None        0       NaN  
23345  dal-vs-bos-0042300405_4_00:22.mp4        1       0.0  

[23346 rows x 32 columns]

mancano 62 video da scaricare

In [34]:
len(df_bos[(df_bos['action_type'] == "Shot") & (df_bos['download'] == 0)])

62

In [35]:
len(df_bos[(df_bos['action_type'] == "Shot") & (df_bos['download'] == 1)])

8944

tra gli 8944 mancano 894 video da cui estrarre i frame

In [37]:
len(df_bos[(df_bos['action_type'] == "Shot") & (df_bos['extracted'] == 0)])

894

In [39]:
len(df_bos[(df_bos['action_type'] == "Shot") & (df_bos['extracted'] == 1)])

8050

video non scaricati + video con frame non trovati + video ok

In [46]:
62+894+8050

9006